astropy/modeling/core.py

```diff
@@ -3341,6 +3341,76 @@ def render(self, out=None, coords=None):
         return out
 
 
+    def replace_submodel(self, name=None, model=None):
+        """
+        Construct a new CompoundModel instance from an existing CompoundModel,
+        replacing the named submodel with the new model.
+
+        In order to ensure that inverses and names are kept/reconstructed, it's
+        necessary to rebuild the CompoundModel from the replaced node all the
+        way back to the base. The original CompoundModel is left untouched.
+
+        Parameters
+        ----------
+        model: CompoundModel instance
+            model to be reconstructed with replacement
+        name: str
+            name of submodel to be replaced
+        new_model: Model
+            replacement model
+        """
+        submodels = [m for m in self.traverse_postorder() if m.name == name]
+        if submodels:
+            if len(submodels) > 1:
+                raise ValueError(f"More than one submodel named {name}")
+
+            # Do this check first in order to raise a more helpful Exception,
+            # although it would fail trying to construct the new CompoundModel
+            old_model = submodels.pop()
+            if (old_model.n_inputs != model.n_inputs or
```

Comment:
Might a check for `n_models` also be needed somewhere here? That's just off the top of my head, from a late-night glance at this, but I see the PR has currently been tagged for `4.1`, so best get any comments in quickly...

Reply:
Yes, I guess it should confirm `len(old_model) == len(model)` although, as noted, if that's not true a `ValueError` gets raised elsewhere. It should work with model sets but I haven't tried it.

I'm also not sure whether `inverse=branch._user_inverse` is the right thing to do. If the new model has an inverse then that won't be propagated up the tree further than a node where a user-defined inverse exists. OTOH, if a user-defined inverse exists for a CompoundModel and this method is being used to replaced one of the submodels, it seems more likely that the user will want to keep that inverse, or they'll redefine it explicitly later. I struggle to see a situation where a user has had to define an inverse for a CompoundModel and then replaces a submodel and suddenly an analytical inverse exists.

Reply:
So there isn't a `ValueError` raised. I can do
```
m = Shift([1, 2], name='m1') | Shift([3, 4], name='m2')
print(m([1, 2]))
mm = m.replace_submodel('m1', Shift(3))
print(mm([1, 2]))
```
and it broadcasts as one might expect. Tickets #9953 and #10170 suggest some behaviour is "accidental" so I've forced a check on `len()` of both models.

Reply:
I believe what you've written above is actually *not* a model set, unless you specify `n_models=2`; it's a model with an array-valued parameter, which is another dimension of complexity... For that reason, your `len` check doesn't actually catch this case, right (though it's still appropriate)?

Reply:
I'll just note in passing that I would actually like to be able to compose a model set with single model and have broadcasting-type behaviour (eg. for wavelength calibrating IFU fibres where each one has a slightly different `Shift` but they all share the same distortion model like a long slit), but that currently doesn't work unless array-valued parameters can be used as above.

Reply:
You're right, I shouldn't have tried to do this last night. But `__len__()` returns `n_models` and that's what's being checked.